File: model/src/main/kotlin/io/sparkled/model/util/IdUtils.kt
package io.sparkled.model.util

import kotlin.random.Random

object IdUtils {

    /**
     * @return A unique 12-digit identifier using time and randomness.
     */
    fun uniqueId(): String {
        val ms = System.currentTimeMillis() * RANDOM_FACTOR
        val rand = Random.nextInt(0, RANDOM_FACTOR)

        return (ms + rand)
            .toString(radix = ID_REMAP_TABLE.size)
            .map { ID_REMAP_TABLE[it] ?: it }
            .joinToString("")
    }

    /**
     * Multiplying the current millis by this factor will yield a 12-digit base 28 identifier. We can then add a random
     * number less than this factor to the timestamp to reduce the possibility of a collision.
     */
    private const val RANDOM_FACTOR = 70000

    private val ID_REMAP_TABLE = mapOf(
        '0' to '2',
        '1' to '3',
        '2' to '4',
        '3' to '5',
        '4' to '6',
        '5' to '7',
        '6' to '8',
        '7' to '9',
        '8' to 'b',
        '9' to 'c',
        'a' to 'd',
        'b' to 'f',
        'c' to 'g',
        'd' to 'h',
        'e' to 'j',
        'f' to 'k',
        'g' to 'm',
        'h' to 'n',
        'i' to 'p',
        'j' to 'q',
        'k' to 'r',
        'l' to 's',
        'm' to 't',
        'n' to 'v',
        'o' to 'w',
        'p' to 'x',
        'q' to 'y',
        'r' to 'z',
    )
}

File: webui/src/utils/idUtils.ts
function uniqueId() {
  const ms = Date.now() * RANDOM_FACTOR
  const rand = Math.random() * RANDOM_FACTOR

  return (ms + rand)
    .toString(Object.keys(ID_REMAP_TABLE).length)
    .split('')
    .map(it => ID_REMAP_TABLE[it] ?? it)
    .join('')
}

/**
 * Multiplying the current millis by this factor will yield a 12-digit base 28 identifier. We can then add a random
 * number less than this factor to the timestamp to reduce the possibility of a collision.
 */
const RANDOM_FACTOR = 70000

const ID_REMAP_TABLE: Record<string, string> = {
  0: '2',
  1: '3',
  2: '4',
  3: '5',
  4: '6',
  5: '7',
  6: '8',
  7: '9',
  8: 'b',
  9: 'c',
  a: 'd',
  b: 'f',
  c: 'g',
  d: 'h',
  e: 'j',
  f: 'k',
  g: 'm',
  h: 'n',
  i: 'p',
  j: 'q',
  k: 'r',
  l: 's',
  m: 't',
  n: 'v',
  o: 'w',
  p: 'x',
  q: 'y',
  r: 'z'
}

export {
  uniqueId
}